src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -249,147 +249,152 @@ private static unsafe void CopyImplPrimitiveWiden(Array sourceArray, int sourceI
-                        {
-                            case CorElementType.ELEMENT_TYPE_I8:
-                                Unsafe.As<byte, long>(ref destElement) = Unsafe.As<byte, int>(ref srcElement); break;
-                            case CorElementType.ELEMENT_TYPE_R4:
-                                Unsafe.As<byte, float>(ref destElement) = Unsafe.As<byte, int>(ref srcElement); break;
-                            case CorElementType.ELEMENT_TYPE_R8:
-                                Unsafe.As<byte, double>(ref destElement) = Unsafe.As<byte, int>(ref srcElement); break;
-                            default:
-                                Debug.Fail("Array.Copy from I4 to another type hit unsupported widening conversion"); break;
-                        }
-                        break;
-
-                    case CorElementType.ELEMENT_TYPE_U8:
-                        switch (destElType)
-                        {
-                            case CorElementType.ELEMENT_TYPE_R4:
-                                Unsafe.As<byte, float>(ref destElement) = Unsafe.As<byte, ulong>(ref srcElement); break;
-                            case CorElementType.ELEMENT_TYPE_R8:
-                                Unsafe.As<byte, double>(ref destElement) = Unsafe.As<byte, ulong>(ref srcElement); break;
-                            default:
-                                Debug.Fail("Array.Copy from U8 to another type hit unsupported widening conversion"); break;
-                        }
-                        break;
-
-                    case CorElementType.ELEMENT_TYPE_I8:
-                        switch (destElType)
-                        {
-                            case CorElementType.ELEMENT_TYPE_R4:
-                                Unsafe.As<byte, float>(ref destElement) = Unsafe.As<byte, long>(ref srcElement); break;
-                            case CorElementType.ELEMENT_TYPE_R8:
-                                Unsafe.As<byte, double>(ref destElement) = Unsafe.As<byte, long>(ref srcElement); break;
-                            default:
-                                Debug.Fail("Array.Copy from I8 to another type hit unsupported widening conversion"); break;
-                        }
-                        break;
-
-                    case CorElementType.ELEMENT_TYPE_R4:
-                        Debug.Assert(destElType == CorElementType.ELEMENT_TYPE_R8);
-                        Unsafe.As<byte, double>(ref destElement) = Unsafe.As<byte, float>(ref srcElement); break;
+                PrimitiveWiden(ref srcElement, ref destElement, srcElType, destElType);
+            }
+        }
 
-                    default:
-                        Debug.Fail("Fell through outer switch in PrimitiveWiden!  Unknown primitive type for source array!"); break;
-                }
+        private static void PrimitiveWiden(ref byte srcElement, ref byte destElement, CorElementType srcElType, CorElementType destElType)
```

Comment:
Does this need `AggressiveInlining` to avoid perf regressions? Native AOT copy of marked with AggressiveInlining.

Reply:
Simple benchmark:

| Method        | Job        | Toolchain              | Mean        | Error     | StdDev    | Median      | Ratio | RatioSD |
|-------------- |----------- |----------------------- |------------:|----------:|----------:|------------:|------:|--------:|
| CopyWiden     | Job-WJOIKQ | \Inline\corerun.exe    |   902.60 ns | 17.893 ns | 44.560 ns |   894.23 ns |  1.00 |    0.00 |
| CopyWiden     | Job-XREDVJ | \NotInline\corerun.exe | 1,553.49 ns | 31.049 ns | 57.551 ns | 1,568.87 ns |  1.73 |    0.08 |
|               |            |                        |             |           |           |             |       |         |
| SetValueWiden | Job-WJOIKQ | \Inline\corerun.exe    |    13.10 ns |  0.285 ns |  0.714 ns |    13.06 ns |  1.00 |    0.00 |
| SetValueWiden | Job-XREDVJ | \NotInline\corerun.exe |    11.25 ns |  0.317 ns |  0.936 ns |    10.65 ns |  0.88 |    0.09 |

`AggressiveInlining` looks better.